libraries/tock-tbf/src/parse.rs

```diff
@@ -147,19 +148,30 @@ pub fn parse_tbf_header(
                         types::TbfHeaderTypes::TbfHeaderMain => {
                             let entry_len = mem::size_of::<types::TbfHeaderV2Main>();
 
-                            // Check that the size of the TLV entry matches the
-                            // size of the Main TLV. If so we can store it.
-                            // Otherwise, we fail to parse this TBF header and
-                            // throw an error.
-                            if tlv_header.length as usize == entry_len {
-                                main_pointer = Some(remaining.try_into()?);
-                            } else {
-                                return Err(types::TbfParseError::BadTlvEntry(
-                                    tlv_header.tipe as usize,
-                                ));
+                            // If there is already a header do nothing: if this is a second Main
+                            // keep the first one, if it's a Program we ignore the Main
+                            if main_pointer.is_none() {
+                                if tlv_header.length as usize == entry_len {
+                                    main_pointer = Some(remaining.try_into()?);
+                                } else {
+                                    return Err(types::TbfParseError::BadTlvEntry(
+                                        tlv_header.tipe as usize,
+                                    ));
+                                }
+                            }
+                        }
+                        types::TbfHeaderTypes::TbfHeaderProgram => {
+                            let entry_len = mem::size_of::<types::TbfHeaderV2Program>();
+                            if program_pointer.is_none() {
```

Comment:
I think we want to be accepting of unexpected TBF headers. Otherwise future updates are more likely to break existing kernels when there otherwise wouldn't be an issue.

Reply:
The TRD says:

```
A TBF object MUST NOT have more than one Program Header. If a TBF
Object has both a Program Header and a Main Header, the kernel's
policy decides which is used. For example, older kernels that do not
understand a Program Header may use the Main Header, while newer
kernels may choose the Program Header.
```

So this isn't something we should be nicely handling

Reply:
> So this isn't something we should be nicely handling

I disagree, but I also didn't write that in the TRD. I'm not sure how we benefit from strictly parsing TBF headers. But I'm also willing to learn more about what the risks are for not rejecting apps that have multiple headers.

Reply:
I think there's a misunderstanding on what a MUST NOT means. MUST NOT doesn't need to be checked; it is an absolute requirement of a TBF object, but not an absolute requirement of a TBF object parser that it throw an error if this MUST NOT is violated.

Reply:
I'm happy not to reject apps, we should then just say that in the documentation. That we use the latest or the first or whatever we end up doing. It seems weird that we say this thing MUST NOT happen, and then have code to handle it in an opaque way (not returning an error, but instead guessing what the user meant)